### ISSUE TITLE
DSL to make FlowReduxStateMachines composable from smaller "sub-statemachines"s

### ISSUE THREAD

**Issue by sockeqwe** [freeletics/FlowRedux#133]
api could looks something like this:

```kotlin
class Myflowreduxstatemachine<MyState MyAction>) {
... 
spec {
   subStateMachine(subStateMachine: FlowReduxStateMachine sm1, actionsMapper : (MyAction) -> SubStateMachineAction,  stateMapper : mySubState -> MyState
}
```

- Mapper functions are not suspending
- do we need to check if statemachine + substatemachines are using same CoroutineScope

**Comment by abhimuktheeswarar** [comment]
Does "composable" refer to defining Hierarchical & Parallel state machines?

**Comment by sockeqwe** [comment]
TL;DR: Assuming that we have the same understanding of the definition of `Hiearchical State Machines` and `Parallel State Machines` then yes, "composable" refers to supporting `Hierarchical & Parallel state machines` (read: can be hierarchical or parallel or both, totally up to the developer).

------- 

#### 1. My definition of Hierchical State Machine:
Let's say we want to show a list of Items but with Pagination support. Then you could have 1 "root" state machine with states `loading first page state`, `error loading first page state`, and `show items state`, and then you could have a "sub state machine that manages while being in  `show items state` the loading of the next page with substates `loading next page state`, `error loading next page state` and `show items with latest loaded page state`.

This would translate to something like that 

```kotlin

class RooStateMachine<State, Action>(nextPageStateMachine : NextPageStateMachine) : FlowReduxStateMachine(initialState = LoadingState) {
   inState<LoadingFirstPageState> {
       onEnter { 
           try {
              val items = loadFirstPageItems()
              OverrideState( ShowItemsState(items) )
       } catch (t : Throwable) {
          OverrideState( ErrorFirstPageState(t) )
       }
   }
   
   inState<ErrorFirstPageState> {
       on<RetryLoadingAction> {
          OverrideState( LoadingFirstPageState() )
       }
   }
   
   // Now instead of having an inState<ShowItemsState> block we  rather hierarchically compose it by splitting this functionality in a "subStateMachine"
   
   subStateMachine(nextPageStateMachine) // register "subStateMachine". It will get all the "SubActions" and emits back "substate" to RooStateMachine
}


// data ShowItemState(val items : List<Item>, nextPageLoadingState : NextPageLoadingState)  with enum class NextPageLoadingState { IDLE, LOADING, ERROR }

class NextPageStateMachine<ShowItemsState, SubActions> : FlowReduxStateMachine(initialState =ShowItemState( ){
  inState <ShowItemsState>(condition = { state : ShowItemsState -> state.nextPageLoadingState == NextPageLoadingState.IDLE}) {
       on<LoadNextPageAction> {
          MutateState { copy( nextPageLoadingState =NextPageLoadingState. LOADING )}
       }
  }
  
  inState <ShowItemsState>(condition = { state : ShowItemsState -> state.nextPageLoadingState == LOADING}) {
      try {
         val nextItems = loadNextPageItems()
         MutateState { copy(items = items + nextItems, nextPageLoadingState = IDLE) }
      catch (t : Throwable) {
         MutateState { copy( nextPageLoadingState = IDLE) }
      }
  }
  
 inState <ShowItemsState>(condition = { state : ShowItemsState -> state.nextPageLoadingState == NextPageLoadingState.ERROR}) {
       delay(3000) // wait 3 seconds, this translates show an error message in UI for 3 seconds
       MutateState {  copy (nextPageLoadingState = IDLE) } 
  }
}
```

#### 2. For parallel state Machines, imho this is just some special case of hierarchical state machines just with indipendent branches.

Let's say we would like to have some sort of screen that has 2 features in one screen: 1. upload Files and 2. download Files
Then you can split this 2 features into parallel state machines like that:

```kotlin

data class ScreenState (val uploadState : UploadState, val downloadState : DownloadState) 

class RootStateMachine <ScreenState, Actions>(uploadStateMachine: UploadStateMachine, downloadStateMachine : DowloadStateMachine) : FlowReduxStateMachine {
  
  spec { 
      subStateMachine(uploadStateMachine, stateMapper = { screenState, uploadState ->  screenState.copy( uploadState = uploadState ) })
      
      subStateMachine(downloadStateMachine, stateMapper = { screenState, downloadState ->  screenState.copy( downloadState = downloadState ) })


    // You could have some more inState{ ... } blocks here below if you want to ...
  }
}
```

---- 

I was actually wondering what a better name for `subStateMachine()` could be to make it more explicit. do you think we should stick to the academic terminology of Hierachical and Parallel State Machines?

**Comment by abhimuktheeswarar** [comment]
We are on the same page on the definition of Hierarchical & Parallel state machines. I'm referring to [XState](https://xstate.js.org/docs/guides/hierarchical.html).   
`subStateMachine` - may not be appropirate for defining parallel state machines. Perhaps we could just use the same or generic name like `stateMachine()`  since the state machine is going to be composbale.
You can refer to W3C SCXML [specification](https://www.w3.org/TR/scxml/ ) for state machines.
